### PR TITLE
fix: ignore masked invalid values in policy loss reductions

### PR DESCRIPTION
**Pull request by haoyang9804**
## Summary

This fixes a quiet PPO training-signal bug in `PolicyLoss.forward`.

`action_mask=0` positions can contain padding/tool/rejected-token values that are semantically ignored by the loss. Before this change, `PolicyLoss` still formed `log_probs - old_log_probs` and later used mask multiplication in reduction paths. If a masked `old_log_probs` slot was `NaN`, both actor loss and `ppo_kl` became `NaN` even though the token was masked out.

The fix zeros masked values before policy-ratio arithmetic and zeros masked per-token losses before `aggregate_loss` reductions. Unmasked invalid values are not hidden.

## Concrete Failure Shape

The smallest failing batch has one valid action token and one masked token:

```text
log_probs:     [[0.0, 0.0]]
old_log_probs: [[0.0, NaN]]
advantages:    [[1.0, 1.0]]
action_mask:   [[1.0, 0.0]]
```

The second token is semantically ignored because `action_mask=0`. It may come from padding, a filtered/rejected rollout position, or another non-trainable response slot. Before this fix, `PolicyLoss.forward` still computed:

```python
raw_policy_log_ratio = log_probs - old_log_probs
```

That produces:

```text
raw_policy_log_ratio: [[0.0, NaN]]
```

Later mask multiplication does not remove the invalid value:

```text
NaN * 0 = NaN
```

So a token that should not contribute to training can still make `loss` and `ppo_kl` non-finite. The fixed path zeroes masked values before ratio/KL/loss arithmetic, so the masked slot stays semantically ignored and the valid token keeps the same gradient.

## Reproduction Recipe

```json
{
  "bug_id": "BUG-4B85D84D7C0E",
  "target": "openrlhf",
  "boundary": "openrlhf.models.loss.PolicyLoss.forward",
  "scenario": {
    "log_probs": [[0.0, 0.0]],
    "old_log_probs": [[0.0, "nan"]],
    "advantages": [[1.0, 1.0]],
    "action_mask": [[1.0, 0.0]]
  },
  "expected_unpatched": {
    "loss_isfinite": false,
    "ppo_kl_isfinite": false
  },
  "expected_fixed": {
    "loss_isfinite": true,
    "ppo_kl_isfinite": true,
    "grad_isfinite": true
  }
}
```

## Portable Runner

```bash
#!/usr/bin/env bash
set -euo pipefail

REPO="${REPO:-$(pwd)}"
export PYTHONPATH="${REPO}${PYTHONPATH:+:${PYTHONPATH}}"

python3 - <<'PY'
import json
import math
import torch

from openrlhf.models.loss import PolicyLoss


def scalar(x):
    value = x.detach().cpu().item()
    if isinstance(value, float) and math.isnan(value):
        return "nan"
    return value


log_probs = torch.nn.Parameter(torch.tensor([[0.0, 0.0]], dtype=torch.float32))
old_log_probs = torch.tensor([[0.0, float("nan")]], dtype=torch.float32)
advantages = torch.tensor([[1.0, 1.0]], dtype=torch.float32)
action_mask = torch.tensor([[1.0, 0.0]], dtype=torch.float32)

loss, clip_ratio, ppo_kl, vllm_kl = PolicyLoss(policy_loss_type="ppo")(
    log_probs=log_probs,
    old_log_probs=old_log_probs,
    advantages=advantages,
    action_mask=action_mask,
)
loss.backward()

print(json.dumps({
    "loss": scalar(loss),
    "loss_isfinite": bool(torch.isfinite(loss).item()),
    "clip_ratio": scalar(clip_ratio),
    "clip_ratio_isfinite": bool(torch.isfinite(clip_ratio).item()),
    "ppo_kl": scalar(ppo_kl),
    "ppo_kl_isfinite": bool(torch.isfinite(ppo_kl).item()),
    "grad": log_probs.grad.detach().cpu().tolist(),
    "grad_isfinite": bool(torch.isfinite(log_probs.grad).all().item()),
    "vllm_kl": vllm_kl,
}, indent=2))
PY
```

## Observed Output

Unpatched OpenRLHF `main@c3188af37cec984614aaa38906e71fa2fc57b079`:

```json
{
  "loss": "nan",
  "loss_isfinite": false,
  "ppo_kl": "nan",
  "ppo_kl_isfinite": false,
  "grad_isfinite": true,
  "update_delta_isfinite": true
}
```

Fixed branch `b99673fa1547a9da8b0762fa5dbb0229eb672b7c`:

```json
{
  "loss": -1.0,
  "loss_isfinite": true,
  "ppo_kl": 0.0,
  "ppo_kl_isfinite": true,
  "grad_isfinite": true,
  "update_delta_isfinite": true
}
```

The fixed validation imports `openrlhf.models.loss` from the repaired checkout and starts the local Ray runtime; vLLM and DeepSpeed imports also pass.

## Duplicate Search

OpenRLHF has no PR template. I followed `CONTRIBUTING.md`, which asks contributors to use pre-commit hooks.

Searched OpenRLHF upstream PRs/issues for exact and related reports:

- `repo:OpenRLHF/OpenRLHF is:pr PolicyLoss old_log_probs NaN action_mask`: 0 results
- `repo:OpenRLHF/OpenRLHF is:pr ppo_kl NaN action_mask`: 0 results
- `repo:OpenRLHF/OpenRLHF is:issue PolicyLoss old_log_probs NaN action_mask`: 0 results

Same-family local RL-infra bugs exist in other projects, but they are not OpenRLHF `PolicyLoss.forward` fixes.

## Tests

```bash
python3 -m pytest tests/test_loss_aggregation.py -q
python3 -m ruff check openrlhf/models/loss.py tests/test_loss_aggregation.py
pre-commit run --files openrlhf/models/loss.py tests/test_loss_aggregation.py
```

Results:

- `tests/test_loss_aggregation.py`: 8 passed
- `ruff`: passed
- `pre-commit`: passed

### CODE REVIEW

**Review by Excelius-Wang — COMMENTED** [review]
Zeroing masked inputs before the ratio arithmetic addresses the part that multiplication by a mask cannot fix: `NaN * 0` is still NaN.

I checked masked NaNs in current log-probabilities, old log-probabilities and advantages for both PPO and GSPO. The loss and diagnostics match the valid-token-only case, and the masked current-log-probability gradient is zero. A NaN at an unmasked position still remains visible. The existing regressions and these additional CPU checks pass. I haven't run distributed training.
